### PR TITLE
Lyra analysis

### PR DESCRIPTION
**Pull request by DanRyanIrish**
Hi guys,
Here is some code that I've developed for the LYRA team to help in the analysis of LYRA data.  It enables python users to easily access information about artifacts and remove them from data.
The principle new functions are:
(1) extract_lytaf_events() which reads information about instrumental effects etc. from the LYRA Time Annotation database files (LYTAF) into a numpy recarray.  This deprecates get_lytaf_events() which does something similar but only accesses the Pointing and Positioning LYTAF database.  extract_lytaf_events() can access all four LYTAF databases.  I've added a Deprecated warning to get_lytaf_events() and also altered split_series_using_lytaf() so that it uses this new function without changing functionality.
(2) remove_lyra_artifacts() which removes periods corresponding to artifacts listed in the LYTAF files from user input time and irradiance arrays. Exactly which artifacts are to be removed can be controlled by the user.
To help with testing, I've also added some simulated LYTAF database files to sunpy/data/test/

I look forward to your comments and suggestions :)

### CODE REVIEW

**Comment by Cadair** [comment]
@DanRyanIrish You have some doc fails: https://travis-ci.org/sunpy/sunpy/jobs/46055593#L1451

**Comment by DanRyanIrish** [comment]
@Cadair I've never been able to understand the Travis error messages.  But I deleted some indentations in a docstring which looks like the problem.  Maybe Travis will pass now...

**Comment by Cadair** [comment]
@DanRyanIrish In theory you can test if sphinx will build correctly by running `python setup.py build_sphinx -W` from the root of the git repo.

File: sunpy/instr/lyra.py [hunk 1]
```diff
@@ -1,123 +1,577 @@
...
+        interesting, e.g. GOES flares.
+
+    References
+    ----------
+    Further documentation: http://proba2.oma.be/data/TARDIS
+
+    Examples
+    --------
+    Get all events in the LYTAF files for January 2014
+        >>> lytaf = get_lytaf_events('2014-01-01', '2014-02-01')
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: I am assuming this is where the error is.

Try formatting it like this:

```python
"""
Get all events in the LYTAF files for January 2014:
     lytaf = get_lytaf_events('2014-01-01', '2014-02-01')
""""

Comment (Cadair) [inline sunpy/instr/lyra.py]: also don't forget a blank line at the end of the indent.

File: sunpy/instr/lyra.py [hunk 2]
```diff
@@ -1,123 +1,581 @@
...
+    ----------
+    lc : LightCurve object
+
+     artifacts : list of strings
+        Contain the artifact types to be removed.  For list of artifact types
+        see reference [1].  For example, if user wants to remove only large
+        angle rotations, listed at reference [1] as LAR, let artifacts=["LAR"].
+        Default=[], i.e. no artifacts will be removed.
+
+    return_artifacts : (optional) bool
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: optional should be removed.

File: sunpy/instr/lyra.py [hunk 3]
```diff
@@ -1,123 +1,581 @@
...
+        cover entire input time range etc.
+        Default=False
+
+    Returns
+    -------
+    lc_new : LYRALightCurve
+        copy of input LYRALightCurve with periods corresponding to artifacts
+        removed.
+
+    artifact_status : (optional) dictionary
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: Should remove optional

File: sunpy/instr/lyra.py [hunk 4]
```diff
@@ -1,123 +1,581 @@
...
+                clean_channels[i] = np.delete(f, bad_indices)
+    # If return_artifacts kwarg is True, return a list containing
+    # information on what artifacts found, removed, etc.  See docstring.
+    if return_artifacts is True:
+        artifact_status = {"lytaf": lytaf,
+                           "removed": lytaf[artifact_indices],
+                           "not_removed": np.delete(lytaf, artifact_indices),
+                           "not_found": artifacts_not_found}
+    # Output FITS file if fits kwarg is set
+    if fitsfile is not None:
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `if fitsfile`

File: sunpy/instr/lyra.py [hunk 5]
```diff
@@ -1,123 +1,581 @@
...
+                cols.append(fits.Column(name=filecolumns[i+1], format="D",
+                                        array=f))
+        coldefs = fits.ColDefs(cols)
+        tbhdu = fits.new_table(coldefs)
+        hdu = fits.PrimaryHDU()
+        tbhdulist = fits.HDUList([hdu, tbhdu])
+        # Write data to fits file.
+        tbhdulist.writeto(fitsfile)
+    # Output csv file if fits kwarg is set.
+    if csvfile is not None:
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `if csvfile`

File: sunpy/instr/lyra.py [hunk 6]
```diff
@@ -1,123 +1,581 @@
...
+                for i in range(len(time)):
+                    csvwriter.writerow(string_time[i])
+            else:
+                for i in range(len(time)):
+                    row = [string_time[i]]
+                    for f in channels:
+                        row.append(f[i])
+                    csvwriter.writerow(row)
+    # Return values.
+    if return_artifacts is True:
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `if return_artifacts`

File: sunpy/instr/lyra.py [hunk 7]
```diff
@@ -1,123 +1,581 @@
 from __future__ import absolute_import
+from __future__ import division
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: These two imports should be one line.

File: sunpy/instr/lyra.py [hunk 8]
```diff
@@ -1,123 +1,581 @@
...
+LYTAF_REMOTE_PATH = "http://proba2.oma.be/lyra/data/lytaf/"
+LYTAF_PATH = config.get("downloads", "download_dir")
 
-    return
+def remove_lytaf_events_from_lightcurve(lc, artifacts=[],
+                                        return_artifacts=False,
+                                        lytaf_path=LYTAF_PATH,
+                                        force_use_local_lytaf=False):
+    """
+    Removes periods of LYRA artifacts defined in LYTAF from a LYRALightCurve.
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `LYRALightCurve` should be a link using the `<backtick>~sunpy.lightcurve.LYRALightCurve<backtick>`

File: sunpy/instr/lyra.py [hunk 9]
```diff
@@ -1,123 +1,581 @@
...
+
+    force_use_local_lytaf : bool
+        Ensures current local version of lytaf files are not replaced by
+        up-to-date online versions even if current local lytaf files do not
+        cover entire input time range etc.
+        Default=False
+
+    Returns
+    -------
+    lc_new : LYRALightCurve
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: should be linked

File: sunpy/instr/lyra.py [hunk 10]
```diff
@@ -1,123 +1,581 @@
...
+        copy of input LYRALightCurve with periods corresponding to artifacts
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: should be linked

File: sunpy/instr/lyra.py [hunk 11]
```diff
@@ -1,123 +1,581 @@
...
+    kwarg, then the relevant values from these arrays are also removed.  This
+    is done by assuming that each element in each array supplied corresponds to
+    the time in the same index in time array.  The artifacts to be removed are
+    given via the artifacts kwarg.  The default is "all", meaning that all
+    artifacts will be removed.  However, a subset of artifacts can be removed
+    by supplying a list of strings of the desired artifact types.
+
+    Parameters
+    ----------
+    time : ndarray/array-like of datetime objects
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: ndarray should be `<backtick>numpy.ndarray<backtick>` and datetime should also be in backticks.

File: sunpy/instr/lyra.py [hunk 12]
```diff
@@ -1,123 +1,581 @@
...
+        Gives the times of the timeseries.
+
+    channels : (optional) list of ndarrays/array-likes convertible to float64.
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: same here.

File: sunpy/instr/lyra.py [hunk 13]
```diff
@@ -1,123 +1,581 @@
...
+        Contains arrays of the irradiances taken at the times in the time
+        variable.  Each element in the list must have the same number of
+        elements as time.
+
+    artifacts : list of strings
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: list should be backticked.

File: sunpy/instr/lyra.py [hunk 14]
```diff
@@ -1,123 +1,581 @@
...
+
+    force_use_local_lytaf : bool
+        Ensures current local version of lytaf files are not replaced by
+        up-to-date online versions even if current local lytaf files do not
+        cover entire input time range etc.
+        Default=False
+
+    Returns
+    -------
+    clean_time : ndarray/array-like of datetime objects
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: backticks.

File: sunpy/instr/lyra.py [hunk 15]
```diff
@@ -1,123 +1,581 @@
...
+                                        eventType_type[id_index],
+                                        eventType_definition[id_index]), dtype=lytaf.dtype))
+        # Close file
+        cursor.close()
+        connection.close()
+    # Sort lytaf in ascending order of begin time
+    np.recarray.sort(lytaf, order="begin_time")
+
+    # If csvfile kwarg is set, write out lytaf to csv file
+    if csvfile is not None:
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `if csvfile`

File: sunpy/instr/lyra.py [hunk 16]
```diff
@@ -1,123 +1,581 @@
...
     """
-    Proba-2 analysis code for splitting up LYRA timeseries around locations where LARs
-    (and other data events) are observed.
+    Proba-2 analysis code for splitting up LYRA timeseries around locations
+    where LARs (and other data events) are observed.
 
     Inputs
     ------
-    timearray - an array of times that can be understood by the SunPy parse_time function
+    timearray - array of times understood by SunPy parse_time function.
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: link the parse_time function

File: sunpy/instr/lyra.py [hunk 17]
```diff
@@ -1,123 +1,581 @@
...
+    return
+
+def split_series_using_lytaf(timearray, data, lytaf):
     """
-    Proba-2 analysis code for splitting up LYRA timeseries around locations where LARs
-    (and other data events) are observed.
+    Proba-2 analysis code for splitting up LYRA timeseries around locations
+    where LARs (and other data events) are observed.
 
     Inputs
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: Should be parameters?

File: sunpy/instr/lyra.py [hunk 18]
```diff
@@ -1,123 +1,581 @@
...
+def remove_lytaf_events_from_lightcurve(lc, artifacts=[],
+                                        return_artifacts=False,
+                                        lytaf_path=LYTAF_PATH,
+                                        force_use_local_lytaf=False):
+    """
+    Removes periods of LYRA artifacts defined in LYTAF from a LYRALightCurve.
+
+    Parameters
+    ----------
+    lc : LightCurve object
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: link Lightcurve

File: sunpy/instr/lyra.py [hunk 19]
```diff
@@ -1,123 +1,581 @@
...
+    if fitsfile is not None:
+        # Create time array of time strings rather than datetime objects
+        # and verify filecolumns have been correctly input.  If None,
+        # generate generic filecolumns (see docstring og function called
+        # below.
+        string_time, filecolumns = _prep_columns(time, channels, filecolumns)
+        # Prepare column objects.
+        cols = [fits.Column(name=filecolumns[0], format="26A",
+                            array=string_time)]
+        if channels is not None:
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `if channels`

File: sunpy/instr/lyra.py [hunk 20]
```diff
@@ -1,123 +1,581 @@
...
+    lc_new.data = pandas.DataFrame(
+        index=time, data=dict((col, channels[i])
+                              for i, col in enumerate(data_columns)))
+    if return_artifacts:
+        return lc_new, artifact_status
+    else:
+        return lc_new
+
+def _remove_lytaf_events(time, channels=None, artifacts=[],
+                         return_artifacts=False, fitsfile=None,
```
Comment (DanRyanIrish) [inline sunpy/instr/lyra.py]: kwargs lytaf_path and artifacts should be None and then defined to default inside function.  Same goes for all other functions.

File: sunpy/instr/lyra.py [hunk 21]
```diff
@@ -1,123 +1,581 @@
...
+                new_row.append(row[2].strftime("%Y-%m-%dT%H:%M:%S"))
+                new_row.append(row[3].strftime("%Y-%m-%dT%H:%M:%S"))
+                new_row.append(row[4])
+                new_row.append(row[5])
+                csvwriter.writerow(new_row)
+
+    return lytaf
+
+def print_lytaf_event_types(lytaf_path=LYTAF_PATH):
+    """Prints the different event types in the each of the LYTAF databases."""
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: This should have a full docstring as it is a public function.

File: sunpy/instr/lyra.py [hunk 22]
```diff
@@ -1,123 +1,581 @@
...
+        cursor = connection.cursor()
+        cursor.execute("select type from eventType;")
+        event_types = cursor.fetchall()
+        print "----------------\n{0} database\n----------------".format(suffix)
+        for event_type in event_types:
+            print str(event_type[0])
+        print " "
+
+def download_lytaf_database(lytaf_dir=''):
+    """download latest Proba2 pointing database from Proba2 Science Center"""
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: Again needs a full docstring (parameters / returns)

**Comment by Cadair** [comment]
I think this is generally very good, just a few relatively minor style and documentation tweaks.

File: sunpy/instr/lyra.py [hunk 23]
```diff
@@ -1,123 +1,625 @@
...
+    kwarg, then the relevant values from these arrays are also removed.  This
+    is done by assuming that each element in each array supplied corresponds to
+    the time in the same index in time array.  The artifacts to be removed are
+    given via the artifacts kwarg.  The default is "all", meaning that all
+    artifacts will be removed.  However, a subset of artifacts can be removed
+    by supplying a list of strings of the desired artifact types.
+
+    Parameters
+    ----------
+    time : `numpy.array` of `datetime.datetime`
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: `numpy.array` should be `numpy.ndarray`

File: sunpy/instr/lyra.py [hunk 24]
```diff
@@ -1,123 +1,625 @@
...
+            for i, f in enumerate(channels):
+                cols.append(fits.Column(name=filecolumns[i+1], format="D",
+                                        array=f))
+        coldefs = fits.ColDefs(cols)
+        tbhdu = fits.new_table(coldefs)
+        hdu = fits.PrimaryHDU()
+        tbhdulist = fits.HDUList([hdu, tbhdu])
+        # Write data to fits file.
+        tbhdulist.writeto(fitsfile)
+    # Output csv file if fits kwarg is set.
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: comment is wrong :p

File: sunpy/instr/lyra.py [hunk 25]
```diff
@@ -1,123 +1,625 @@
...
+        else:
+            return clean_time, clean_channels, artifact_status
+    else:
+        if not channels:
+            return clean_time
+        else:
+            return clean_time, clean_channels
+
+def get_lytaf_events(start_time, end_time, lytaf_path=None,
+                     combine_files=["lyra", "manual", "ppt", "science"],
```
Comment (Cadair) [inline sunpy/instr/lyra.py]: I would make this combine_files a tuple here to protect it from trouble later on.

**Comment by Cadair** [comment]
Alright @DanRyanIrish you may have my very hard earned :+1:

**Comment by DanRyanIrish** [comment]
Thanks God for that! ;)

**Comment by Cadair** [comment]
@ayshih @aringlis @dpshelio @ehsteve We should merge this later, please can someone do some kind of review even if not in painstaking detail.

**Comment by Cadair** [comment]
bite me coveralls.

**Comment by ehsteve** [comment]
:+1:

**Pull request merged**